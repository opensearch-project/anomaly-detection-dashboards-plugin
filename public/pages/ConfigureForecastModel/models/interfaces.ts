/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

// Formik values used when creating the model configuration
export interface ModelConfigurationFormikValues {
  shingleSize: number;
  imputationOption?: ImputationFormikValues;
  interval: number | undefined;
  windowDelay: number | undefined;
  suggestedSeasonality?: number;
  recencyEmphasis?: number;
  resultIndexMinAge?: number | string;
  resultIndexMinSize?: number | string;
  resultIndexTtl?:number | string;
  flattenCustomResultIndex?: boolean;
  resultIndex?: string;
  horizon?: number;
  history?: number;
}

export interface ImputationFormikValues {
  imputationMethod?: string;
  custom_value?: CustomValueFormikValues[];
}

export interface CustomValueFormikValues {
  // forecasting has only one feature and we don't need to specify the feature name
  featureName?: string;
  data: number;
}
