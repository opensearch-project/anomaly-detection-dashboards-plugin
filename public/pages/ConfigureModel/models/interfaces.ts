/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import { AggregationOption } from '../../../models/types';

// Formik values used when creating/editing the model configuration
export interface ModelConfigurationFormikValues {
  featureList: FeaturesFormikValues[];
  categoryFieldEnabled: boolean;
  categoryField: string[];
  shingleSize: number;
  imputationOption?: ImputationFormikValues;
  suppressionRules?: RuleFormikValues[];
}

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string | undefined;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}

export interface ImputationFormikValues {
  imputationMethod?: string;
  custom_value?: CustomValueFormikValues[];
}

export interface CustomValueFormikValues {
  featureName: string;
  data: number;
}

export interface RuleFormikValues {
  featureName: string;
  absoluteThreshold?: number;
  relativeThreshold?: number;
  aboveBelow: string;
}
