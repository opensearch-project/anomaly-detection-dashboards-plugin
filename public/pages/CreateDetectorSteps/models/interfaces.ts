/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';
import { ModelConfigurationFormikValues } from '../../ConfigureModel/models/interfaces';
import { DetectorJobsFormikValues } from '../../DetectorJobs/models/interfaces';

// Formik values used upon creation (includes all fields + those related to historical detector date range)
export interface CreateDetectorFormikValues
  extends DetectorDefinitionFormikValues,
    ModelConfigurationFormikValues,
    DetectorJobsFormikValues {}
