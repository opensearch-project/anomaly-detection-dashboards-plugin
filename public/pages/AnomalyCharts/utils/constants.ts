/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

export const INITIAL_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0.0,
  maxAnomalyGrade: 0.0,
  minConfidence: 0.0,
  maxConfidence: 0.0,
  lastAnomalyOccurrence: '',
};

export enum CHART_FIELDS {
  PLOT_TIME = 'plotTime',
  ANOMALY_GRADE = 'anomalyGrade',
  CONFIDENCE = 'confidence',
  DATA = 'data',
  AGG_INTERVAL = 'aggInterval',
  EXPECTED_VALUE = 'expectedValue',
}

export enum CHART_COLORS {
  ANOMALY_GRADE_COLOR = '#D13212',
  FEATURE_DATA_COLOR = '#16191F',
  FEATURE_COLOR = '#fcd529',
  CONFIDENCE_COLOR = '#017F75',
  LIGHT_BACKGROUND = '#FFFFFF',
  DARK_BACKGROUND = '#1D1E24',
}

export const FEATURE_CHART_THEME = [
  {
    lineSeriesStyle: {
      line: {
        strokeWidth: 2,
        visible: true,
        opacity: 0.5,
      },
      point: {
        visible: true,
        stroke: CHART_COLORS.FEATURE_DATA_COLOR,
      },
    },
  },
  {
    colors: {
      vizColors: [CHART_COLORS.FEATURE_DATA_COLOR],
    },
  },
];

export const ANOMALY_CHART_THEME = [
  {
    colors: {
      vizColors: [
        CHART_COLORS.CONFIDENCE_COLOR,
        CHART_COLORS.ANOMALY_GRADE_COLOR,
      ],
    },
  },
];

export const DATE_PICKER_QUICK_OPTIONS = [
  { start: 'now-24h', end: 'now', label: 'last 24 hours' },
  { start: 'now-7d', end: 'now', label: 'last 7 days' },
  { start: 'now-30d', end: 'now', label: 'last 30 days' },
  { start: 'now-90d', end: 'now', label: 'last 90 days' },

  { start: 'now/d', end: 'now', label: 'Today' },
  { start: 'now/w', end: 'now', label: 'Week to date' },
  { start: 'now/M', end: 'now', label: 'Month to date' },
  { start: 'now/y', end: 'now', label: 'Year to date' },
];

export enum LIVE_CHART_CONFIG {
  REFRESH_INTERVAL_IN_SECONDS = 30 * 1000, //poll anomaly result every 30 seconds
  MONITORING_INTERVALS = 60,
}

export const DEFAULT_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0,
  maxAnomalyGrade: 0,
  minConfidence: 0,
  maxConfidence: 0,
  lastAnomalyOccurrence: '-',
};

export const HEATMAP_CHART_Y_AXIS_WIDTH = 30;
