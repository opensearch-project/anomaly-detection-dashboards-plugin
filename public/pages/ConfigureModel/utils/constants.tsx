/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import {
  ModelConfigurationFormikValues,
  FeaturesFormikValues,
} from '../../ConfigureModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    featureList: [],
    categoryFieldEnabled: false,
    categoryField: [],
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined
  };

export const INITIAL_FEATURE_VALUES: FeaturesFormikValues = {
  featureId: '',
  featureName: '',
  featureEnabled: true,
  featureType: FEATURE_TYPE.SIMPLE,
  aggregationQuery: JSON.stringify(
    { aggregation_name: { sum: { field: 'field_name' } } },
    null,
    4
  ),
  aggregationBy: '',
  aggregationOf: [],
};

export const FEATURE_TYPES = [
  { text: 'Custom Aggregation', value: FEATURE_TYPE.CUSTOM },
  { text: 'Defined Aggregation', value: FEATURE_TYPE.SIMPLE },
];

export const FEATURE_TYPE_OPTIONS = [
  { text: 'Field value', value: FEATURE_TYPE.SIMPLE },
  { text: 'Custom expression', value: FEATURE_TYPE.CUSTOM },
];

export enum SAVE_FEATURE_OPTIONS {
  START_AD_JOB = 'start_ad_job',
  KEEP_AD_JOB_STOPPED = 'keep_ad_job_stopped',
}

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'value_count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const FEATURE_FIELDS = [
  'featureName',
  'aggregationOf',
  'aggregationBy',
  'aggregationQuery',
];

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
