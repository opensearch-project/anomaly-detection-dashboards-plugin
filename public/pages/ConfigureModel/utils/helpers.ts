/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { DATA_TYPES, DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';
import {
  FEATURE_TYPE,
  FeatureAttributes,
  Detector,
} from '../../../models/interfaces';
import { v4 as uuidv4 } from 'uuid';
import { get, forOwn, cloneDeep, isEmpty } from 'lodash';
import { DataTypes } from '../../../redux/reducers/opensearch';
import {
  ModelConfigurationFormikValues,
  FeaturesFormikValues,
  CustomValueFormikValues,
  ImputationFormikValues,
  RuleFormikValues,
} from '../../ConfigureModel/models/interfaces';
import { INITIAL_MODEL_CONFIGURATION_VALUES } from '../../ConfigureModel/utils/constants';
import {
  featuresToUIMetadata,
  formikToFeatureAttributes,
} from '../../ReviewAndCreate/utils/helpers';
import {
  ImputationMethod,
  ImputationOption,
  Condition,
  Rule,
  ThresholdType,
  Operator,
  Action,
} from '../../../models/types';
import {
  SparseDataOptionValue
} from './constants'

export const getFieldOptions = (
  allFields: { [key: string]: string[] },
  validTypes: DATA_TYPES[]
) =>
  validTypes
    .map((dataType) =>
      allFields[dataType]
        ? {
            label: dataType,
            options: allFields[dataType].map((field) => ({
              label: field,
              type: dataType,
            })),
          }
        : []
    )
    .filter(Boolean);

export const getNumberFieldOptions = (allFields: { [key: string]: string[] }) =>
  getFieldOptions(allFields, [DATA_TYPES.NUMBER]);

export const getCountableFieldOptions = (allFields: {
  [key: string]: string[];
}) => {
  const countableDataTypes = [
    DATA_TYPES.NUMBER,
    DATA_TYPES.BOOLEAN,
    DATA_TYPES.KEYWORD,
    DATA_TYPES.DATE,
  ];
  return getFieldOptions(
    allFields,
    Object.keys(allFields)
      .map((field) => field as DATA_TYPES)
      .filter((field) => countableDataTypes.includes(field))
  );
};

export const initialFeatureValue = () => ({
  featureId: uuidv4(),
  featureName: undefined,
  featureType: FEATURE_TYPE.SIMPLE,
  featureEnabled: true,
  importance: 1,
  aggregationBy: 'sum',
  aggregationQuery: JSON.stringify(
    {
      aggregation_name: { sum: { field: 'field_name' } },
    },
    null,
    4
  ),
  newFeature: true,
});

export const validateFeatures = (values: any) => {
  const featureList = get(values, 'featureList', []);
  let featureNameCount = new Map<string, number>();

  featureList.forEach((attribute: FeatureAttributes) => {
    if (attribute.featureName) {
      const featureName = attribute.featureName.toLowerCase();
      if (featureNameCount.has(featureName)) {
        featureNameCount.set(
          featureName,
          // @ts-ignore
          featureNameCount.get(featureName) + 1
        );
      } else {
        featureNameCount.set(featureName, 1);
      }
    }
  });

  let hasError = false;
  const featureErrors = featureList.map((attribute: FeatureAttributes) => {
    if (attribute.featureName) {
      // @ts-ignore
      if (featureNameCount.get(attribute.featureName.toLowerCase()) > 1) {
        hasError = true;
        return { featureName: 'Duplicate feature name' };
      } else {
        return undefined;
      }
    } else {
      hasError = true;
      // @ts-ignore
      return {
        featureName: 'You must enter a feature name',
      };
    }
  });

  return hasError ? { featureList: featureErrors } : undefined;
};

export const generateInitialFeatures = (
  detector: Detector
): FeaturesFormikValues[] => {
  const featureUiMetaData = get(detector, 'uiMetadata.features', []);
  const features = get(detector, 'featureAttributes', []);
  // @ts-ignore
  return features.map((feature: FeatureAttributes) => {
    return {
      ...featureUiMetaData[feature.featureName],
      ...feature,
      aggregationQuery: JSON.stringify(feature['aggregationQuery'], null, 4),
      aggregationOf: get(
        featureUiMetaData,
        `${feature.featureName}.aggregationOf`
      )
        ? [
            {
              label: get(
                featureUiMetaData,
                `${feature.featureName}.aggregationOf`
              ),
            },
          ]
        : [],
      featureType: get(featureUiMetaData, `${feature.featureName}.featureType`)
        ? get(featureUiMetaData, `${feature.featureName}.featureType`)
        : FEATURE_TYPE.CUSTOM,
    };
  });
};

export const focusOnFirstWrongFeature = (errors: any, setFieldTouched: any) => {
  if (
    //@ts-ignore
    !!get(errors, 'featureList', []).filter((featureError) => featureError)
      .length
  ) {
    const featureList = get(errors, 'featureList', []);
    for (let i = featureList.length - 1; i >= 0; i--) {
      if (featureList[i]) {
        forOwn(featureList[i], function (value, key) {
          setFieldTouched(`featureList.${i}.${key}`, true);
        });
        focusOnFeatureAccordion(i);
      }
    }
    return true;
  }
  return false;
};

export const focusOnFeatureAccordion = (index: number) => {
  const featureAccordion = document.getElementById(
    `featureAccordionHeaders.${index}`
  );
  //@ts-ignore
  featureAccordion.setAttribute('tabindex', '-1');
  //@ts-ignore
  featureAccordion.focus();
  const header =
    //@ts-ignore
    featureAccordion.parentElement.parentElement.parentElement.parentElement;
  //@ts-ignore
  if (!header.className.includes('euiAccordion-isOpen')) {
    //@ts-ignore
    featureAccordion.click();
  }
};

export const focusOnCategoryField = () => {
  const component = document.getElementById('categoryFieldCheckbox');
  component?.focus();
};

export const getCategoryFields = (dataTypes: DataTypes) => {
  const keywordFields = get(dataTypes, 'keyword', []);
  const ipFields = get(dataTypes, 'ip', []);
  return keywordFields.concat(ipFields);
};

export const focusOnImputationOption = () => {
  const component = document.getElementById('imputationOption');
  component?.focus();
};

export const focusOnSuppressionRules = () => {
  const component = document.getElementById('suppressionRules');
  component?.focus();
};

export const getShingleSizeFromObject = (obj: object) => {
  return get(obj, 'shingleSize', DEFAULT_SHINGLE_SIZE);
};

export function clearModelConfiguration(ad: Detector): Detector {
  return {
    ...ad,
    featureAttributes: [],
    uiMetadata: {
      ...ad.uiMetadata,
      features: {},
    },
    categoryField: undefined,
    shingleSize: DEFAULT_SHINGLE_SIZE,
  };
}

export function modelConfigurationToFormik(
  detector: Detector
): ModelConfigurationFormikValues {
  const initialValues = cloneDeep(INITIAL_MODEL_CONFIGURATION_VALUES);
  if (isEmpty(detector)) {
    return initialValues;
  }

  var imputationMethod = imputationMethodToFormik(detector);

  var defaultFillArray: CustomValueFormikValues[] = [];

  if (SparseDataOptionValue.CUSTOM_VALUE === imputationMethod) {
    const defaultFill = get(detector, 'imputationOption.defaultFill', null) as Array<{ featureName: string; data: number }> | null;
    defaultFillArray = defaultFill
    ? defaultFill.map(({ featureName, data }) => ({
        featureName,
        data,
      }))
    : [];
  }

  const imputationFormikValues: ImputationFormikValues = {
    imputationMethod: imputationMethod,
    custom_value: SparseDataOptionValue.CUSTOM_VALUE === imputationMethod ? defaultFillArray : undefined,
  };

  return {
    ...initialValues,
    featureList: featuresToFormik(detector),
    categoryFieldEnabled: !isEmpty(get(detector, 'categoryField', [])),
    categoryField: get(detector, 'categoryField', []),
    shingleSize: get(detector, 'shingleSize', DEFAULT_SHINGLE_SIZE),
    imputationOption: imputationFormikValues,
    suppressionRules: rulesToFormik(detector.rules),
  };
}

function featuresToFormik(detector: Detector): FeaturesFormikValues[] {
  const featureUiMetaData = get(detector, 'uiMetadata.features', []);
  const features = get(detector, 'featureAttributes', []);
  // @ts-ignore
  return features.map((feature: FeatureAttributes) => {
    return {
      ...featureUiMetaData[feature.featureName],
      ...feature,
      aggregationQuery: JSON.stringify(feature['aggregationQuery'], null, 4),
      aggregationOf: get(
        featureUiMetaData,
        `${feature.featureName}.aggregationOf`
      )
        ? [
            {
              label: get(
                featureUiMetaData,
                `${feature.featureName}.aggregationOf`
              ),
            },
          ]
        : [],
      featureType: get(featureUiMetaData, `${feature.featureName}.featureType`)
        ? get(featureUiMetaData, `${feature.featureName}.featureType`)
        : FEATURE_TYPE.CUSTOM,
    };
  });
}

export function formikToModelConfiguration(
  values: ModelConfigurationFormikValues,
  detector: Detector
): Detector {
  let detectorBody = {
    ...detector,
    uiMetadata: {
      ...detector.uiMetadata,
      features: { ...featuresToUIMetadata(values.featureList) },
    },
    featureAttributes: formikToFeatureAttributes(values.featureList),
    shingleSize: values.shingleSize,
    categoryField: !isEmpty(values?.categoryField)
      ? values.categoryField
      : undefined,
    imputationOption: formikToImputationOption(values.imputationOption),
    rules: formikToRules(values.suppressionRules),
  } as Detector;

  return detectorBody;
}

export function prepareDetector(
  featureValues: FeaturesFormikValues[],
  shingleSizeValue: number,
  categoryFields: string[],
  ad: Detector,
  forPreview: boolean = false
): Detector {
  const detector = cloneDeep(ad);
  const featureAttributes = formikToFeatures(featureValues, forPreview);

  return {
    ...detector,
    featureAttributes: [...featureAttributes],
    shingleSize: shingleSizeValue,
    categoryField: isEmpty(categoryFields) ? undefined : categoryFields,
    uiMetadata: {
      ...detector.uiMetadata,
      features: { ...featuresToUIMetadata(featureValues) },
    },
  };
}

function formikToFeatures(values: FeaturesFormikValues[], forPreview: boolean) {
  const featureAttribute = formikToFeatureAttributes(values, forPreview);
  return featureAttribute;
}

export function formikToSimpleAggregation(value: FeaturesFormikValues) {
  if (
    value.aggregationBy &&
    value.aggregationOf &&
    value.aggregationOf.length > 0
  ) {
    return {
      [value.featureName]: {
        [value.aggregationBy]: { field: value.aggregationOf[0].label },
      },
    };
  } else {
    return {};
  }
}

export function formikToImputationOption(imputationFormikValues?: ImputationFormikValues): ImputationOption | undefined {
  // Map the formik method to the imputation method; return undefined if method is not recognized.
  const method = formikToImputationMethod(imputationFormikValues?.imputationMethod);
  if (!method) return undefined;

  // Convert custom_value array to defaultFill if the method is FIXED_VALUES.
  const defaultFill = method === ImputationMethod.FIXED_VALUES
    ? imputationFormikValues?.custom_value?.map(({ featureName, data }) => ({
        featureName,
        data,
      }))
    : undefined;

  // Construct and return the ImputationOption object.
  return { method, defaultFill };
}

export function imputationMethodToFormik(
  detector: Detector
): string {
  var imputationMethod = get(detector, 'imputationOption.method', undefined) as ImputationMethod;

  switch (imputationMethod) {
    case ImputationMethod.FIXED_VALUES:
      return SparseDataOptionValue.CUSTOM_VALUE;
    case ImputationMethod.PREVIOUS:
      return SparseDataOptionValue.PREVIOUS_VALUE;
    case ImputationMethod.ZERO:
      return SparseDataOptionValue.SET_TO_ZERO;
    default:
      break;
  }

  return SparseDataOptionValue.IGNORE;
}

export function formikToImputationMethod(
  formikValue: string | undefined
): ImputationMethod | undefined {
  switch (formikValue) {
    case SparseDataOptionValue.CUSTOM_VALUE:
      return ImputationMethod.FIXED_VALUES;
    case SparseDataOptionValue.PREVIOUS_VALUE:
      return ImputationMethod.PREVIOUS;
    case SparseDataOptionValue.SET_TO_ZERO:
      return ImputationMethod.ZERO;
    default:
      return undefined;
  }
}

export const getCustomValueStrArray = (imputationMethodStr : string, detector: Detector): string[] => {
  if (SparseDataOptionValue.CUSTOM_VALUE === imputationMethodStr) {
    const defaultFill : Array<{ featureName: string; data: number }> = get(detector, 'imputationOption.defaultFill', []);

    return defaultFill
      .map(({ featureName, data }) => `${featureName}: ${data}`);
  }
  return []
}

export const getSuppressionRulesArray = (detector: Detector): string[] => {
  if (!detector.rules || detector.rules.length === 0) {
    return []; // Return an empty array if there are no rules
  }

  return detector.rules.flatMap((rule) => {
    // Convert each condition to a readable string
    return rule.conditions.map((condition) => {
      const featureName = condition.featureName;
      const thresholdType = condition.thresholdType;
      let value = condition.value;
      const isPercentage = thresholdType === ThresholdType.ACTUAL_OVER_EXPECTED_RATIO || thresholdType === ThresholdType.EXPECTED_OVER_ACTUAL_RATIO;

      // If it is a percentage, multiply by 100
      if (isPercentage) {
        value *= 100;
      }

      // Determine whether it is "above" or "below" based on ThresholdType
      const aboveOrBelow = thresholdType === ThresholdType.ACTUAL_OVER_EXPECTED_MARGIN || thresholdType === ThresholdType.ACTUAL_OVER_EXPECTED_RATIO ? 'above' : 'below';

      // Construct the formatted string
      return `Ignore anomalies for feature "${featureName}" with no more than ${value}${isPercentage ? '%' : ''} ${aboveOrBelow} expected value.`;
    });
  });
};


// Convert RuleFormikValues[] to Rule[]
export const formikToRules = (formikValues?: RuleFormikValues[]): Rule[] | undefined => {
  if (!formikValues || formikValues.length === 0) {
    return undefined; // Return undefined for undefined or empty input
  }

  return formikValues.map((formikValue) => {
    const conditions: Condition[] = [];

    // Determine the threshold type based on aboveBelow and the threshold type (absolute or relative)
    const getThresholdType = (aboveBelow: string, isAbsolute: boolean): ThresholdType => {
      if (isAbsolute) {
        return aboveBelow === 'above'
          ? ThresholdType.ACTUAL_OVER_EXPECTED_MARGIN
          : ThresholdType.EXPECTED_OVER_ACTUAL_MARGIN;
      } else {
        return aboveBelow === 'above'
          ? ThresholdType.ACTUAL_OVER_EXPECTED_RATIO
          : ThresholdType.EXPECTED_OVER_ACTUAL_RATIO;
      }
    };

    // Check if absoluteThreshold is provided, create a condition
    if (formikValue.absoluteThreshold !== undefined && formikValue.absoluteThreshold !== 0 && formikValue.absoluteThreshold !== null
      && typeof formikValue.absoluteThreshold === 'number' && // Check if it's a number
      !isNaN(formikValue.absoluteThreshold) && // Ensure it's not NaN
      formikValue.absoluteThreshold > 0 // Check if it's positive
    ) {
      conditions.push({
        featureName: formikValue.featureName,
        thresholdType: getThresholdType(formikValue.aboveBelow, true),
        operator: Operator.LTE,
        value: formikValue.absoluteThreshold,
      });
    }

    // Check if relativeThreshold is provided, create a condition
    if (formikValue.relativeThreshold !== undefined && formikValue.relativeThreshold !== 0 && formikValue.relativeThreshold !== null
      && typeof formikValue.relativeThreshold === 'number' && // Check if it's a number
      !isNaN(formikValue.relativeThreshold) && // Ensure it's not NaN
      formikValue.relativeThreshold > 0 // Check if it's positive
    ) {
      conditions.push({
        featureName: formikValue.featureName,
        thresholdType: getThresholdType(formikValue.aboveBelow, false),
        operator: Operator.LTE,
        value: formikValue.relativeThreshold / 100,  // Convert percentage to decimal,
      });
    }

    return {
      action: Action.IGNORE_ANOMALY,
      conditions,
    };
  });
};

// Convert Rule[] to RuleFormikValues[]
export const rulesToFormik = (rules?: Rule[]): RuleFormikValues[] => {
  if (!rules || rules.length === 0) {
    return []; // Return empty array for undefined or empty input
  }

  return rules.map((rule) => {
    // Start with default values
    const formikValue: RuleFormikValues = {
      featureName: '',
      absoluteThreshold: undefined,
      relativeThreshold: undefined,
      aboveBelow: 'above', // Default to 'above', adjust as needed
    };

    // Loop through conditions to populate formikValue
    rule.conditions.forEach((condition) => {
      formikValue.featureName = condition.featureName;

      // Determine the value and type of threshold
      switch (condition.thresholdType) {
        case ThresholdType.ACTUAL_OVER_EXPECTED_MARGIN:
          formikValue.absoluteThreshold = condition.value;
          formikValue.aboveBelow = 'above';
          break;
        case ThresholdType.EXPECTED_OVER_ACTUAL_MARGIN:
          formikValue.absoluteThreshold = condition.value;
          formikValue.aboveBelow = 'below';
          break;
        case ThresholdType.ACTUAL_OVER_EXPECTED_RATIO:
          // *100 to convert to percentage
          formikValue.relativeThreshold = condition.value * 100;
          formikValue.aboveBelow = 'above';
          break;
        case ThresholdType.EXPECTED_OVER_ACTUAL_RATIO:
          // *100 to convert to percentage
          formikValue.relativeThreshold = condition.value * 100;
          formikValue.aboveBelow = 'below';
          break;
        default:
          break;
      }
    });

    return formikValue;
  });
};


