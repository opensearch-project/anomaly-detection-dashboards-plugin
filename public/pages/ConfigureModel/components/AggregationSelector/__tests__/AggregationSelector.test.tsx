/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Formik } from 'formik';
import { AggregationSelector } from '../AggregationSelector';
import {
  initialState,
  mockedStore,
} from '../../../../../redux/utils/testUtils';
import { FeaturesFormikValues } from '../../../models/interfaces';
import { INITIAL_FEATURE_VALUES } from '../../../utils/constants';

const renderAggregationSelector = (initialValue: FeaturesFormikValues) => ({
  ...render(
    <Provider
      store={mockedStore({
        ...initialState,
        opensearch: {
          ...initialState.opensearch,
          dataTypes: {
            keyword: ['cityName.keyword'],
            integer: ['age'],
            text: ['cityName'],
          },
        },
      })}
    >
      <Formik initialValues={initialValue} onSubmit={jest.fn()}>
        {(formikProps) => (
          <div>
            <AggregationSelector />
          </div>
        )}
      </Formik>
    </Provider>
  ),
});

describe('<AggregationSelector /> spec', () => {
  describe('Empty results', () => {
    test('renders component with aggregation types and defaults to empty', () => {
      const { container } = renderAggregationSelector(INITIAL_FEATURE_VALUES);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
