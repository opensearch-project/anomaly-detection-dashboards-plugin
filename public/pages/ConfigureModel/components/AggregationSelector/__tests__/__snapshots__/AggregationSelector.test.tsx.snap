// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AggregationSelector /> spec Empty results renders component with aggregation types and defaults to empty 1`] = `
<div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        Aggregation method
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <select
            aria-describedby="random_html_id-help-0"
            class="euiSelect"
            data-test-subj="aggregationType"
            id="random_html_id"
            name="featureList.undefined.aggregationBy"
          >
            <option
              value="avg"
            >
              average()
            </option>
            <option
              value="value_count"
            >
              count()
            </option>
            <option
              value="sum"
            >
              sum()
            </option>
            <option
              value="min"
            >
              min()
            </option>
            <option
              value="max"
            >
              max()
            </option>
          </select>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        class="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        The aggregation method determines what constitutes an anomaly. For example, if you choose min(), the detector focuses on finding anomalies based on the minimum values of your feature.
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        Field
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="euiComboBox"
        data-test-subj="featureFieldTextInput-undefined"
        name="featureList.undefined.aggregationOf"
        role="combobox"
      >
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <div
              class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
              data-test-subj="comboBoxInput"
              tabindex="-1"
            >
              <p
                class="euiComboBoxPlaceholder"
              >
                Select field
              </p>
              <div
                class="euiComboBox__input"
                style="font-size: 14px; display: inline-block;"
              >
                <input
                  aria-controls=""
                  data-test-subj="comboBoxSearchInput"
                  id="random_html_id"
                  role="textbox"
                  style="box-sizing: content-box; width: 2px;"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
            <div
              class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
            >
              <button
                aria-label="Open list of options"
                class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                data-test-subj="comboBoxToggleListButton"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                  focusable="false"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
